core/types/calls.cc

```diff
@@ -1916,6 +1916,45 @@ class Array_flatten : public IntrinsicMethod {
     }
 } Array_flatten;
 
+class Array_product : public IntrinsicMethod {
+public:
+    void apply(Context ctx, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        // Unwrap the array one time to get the element type (we'll rewrap it down at the bottom)
+        TypePtr element;
+        if (auto *ap = cast_type<AppliedType>(thisType)) {
+            ENFORCE(ap->klass == Symbols::Array() || ap->klass.data(ctx)->derivesFrom(ctx, Symbols::Array()));
+            ENFORCE(!ap->targs.empty());
+            element = ap->targs.front();
+        } else if (auto *tuple = cast_type<TupleType>(thisType)) {
+            element = tuple->elementType();
+        } else {
+            ENFORCE(false, "Array#product on unexpected type: {}", args.selfType->show(ctx));
```

Comment:
if this goes through, this should be a user-visible error.
Otherwise, this will be a user-visible segfault.

Reply:
I actually think this one is fine to be an enforce, because this is on the `thisType` which since this intrinsic is on `Array#product`, we'll know that this `thisType` is always either an AppliedType of Array or a TupleType.

tl;dr, i think this ENFORCE is fine (especially because we have nearly the exact same check in `Array#flatten` and it has never fired)